### PR TITLE
[NA] [SDK] Evaluation suite API improvements and feedback fixes

### PR DESCRIPTION
**Pull request by alexkuzmik**
## Summary

Address feedback on the Evaluation Suite SDK API: simplify the public surface, add missing features, improve result display, enforce strict input validation, and optimize scoring performance.

### 1. Hide evaluator concept from public API
- Replace all user-facing `evaluators`/`LLMJudge` references with `assertions: List[str]`
- `get_evaluators()` → `get_assertions()` returning `List[str]`
- `get_items()` returns `"assertions": List[str]` instead of `"evaluators"`
- Remove `evaluators` param from `update()`, `add_item()`, `create/get_or_create`
- Rename `evaluator_model` → `model` in `run()`

### 2. Fix `get_or_create_evaluation_suite` to update existing suites
- Now calls `update()` when assertions or execution_policy are provided for an existing suite
- `update()` supports partial updates: omitted params retain current values

### 3. Task return value validation
- Wrap the task function with validation that checks it returns a dict with `input` and `output` keys
- Clear error messages with examples when the format is wrong
- Uses `functools.wraps` to preserve the original function name for traces/spans

### 4. Tags support
- Add `tags` parameter to `create_evaluation_suite()`, `get_or_create_evaluation_suite()`, and `suite.update()`
- Add `get_tags()` to both `Dataset` and `EvaluationSuite`

### 5. Suite-specific result display with pass rates
- Replace avg score display with pass rate view for evaluation suites
- Categorize assertions into: critical (caused item failures, red), non-critical (items still passed due to execution_policy, yellow), and fully passing (green)
- Display shows items passed count and pass rate separately

### 6. Batch `add_items()` and strict input validation
- Add `add_items(items: List[EvaluationSuiteItem])` for batch item insertion (single dataset version instead of N versions)
- Add `EvaluationSuiteItem` TypedDict for type hints
- Add `validate_execution_policy()` and `validate_suite_items()` with strict checks for unknown keys, types, and required fields
- Validation applied everywhere: `add_item`, `add_items`, `update`, `create`, and `get_or_create`

### 7. Public exports and linter fixes
- Export `EvaluationSuite` and `EvaluationSuiteResult` from `opik.__init__`
- Mark `EvaluationSuite.__init__` as internal (users should use `Opik()` factory methods)
- Fix mypy and ruff warnings

### 8. PR review feedback fixes
- Fix `pass_threshold=0` being coerced to 1 via `or` operator — now uses explicit `is not None` check
- Fix `ZeroDivisionError` in assertion sort lambda when assertion has 0 total count
- Fix `verbose=0` still printing experiment URL — moved inside `if verbose >= 1` block
- Replace bare `assert` for None `version_info` with descriptive `RuntimeError`
- Extract `validate_task_result()` as module-level function for unit testability (removed expensive e2e validation tests)

### 9. Handle 0 items edge case
- `EvaluationSuiteResult.pass_rate` returns `Optional[float]` — `None` when no items evaluated
- Suite report shows yellow "NO ITEMS" and "N/A" pass rate for empty suites

### 10. Per-item timing in report
- Add `task_execution_time` and `scoring_time` optional fields to `TestResult`
- Capture timing in `_compute_test_result_for_llm_task()` using `time.perf_counter()`
- Display avg task time, avg scoring time, and avg total time in suite report (always shown when available)

### 11. Merge suite + item LLM judges (perf optimization)
- Add `LLMJudge.merged()` factory method that combines assertions from multiple judges into one
- `build_metrics_evaluator()` now merges suite-level and item-level LLMJudge instances, reducing sequential LLM API calls from N to 1 per item
- Deduplicates assertions while preserving order

### 12. Deduplicate pass/fail logic, return EvaluationSuiteResult from evaluate_suite
- Remove duplicated `_compute_item_passed_map` from `report.py`, reuse `EvaluationSuiteResult`
- `evaluate_suite` now returns `EvaluationSuiteResult` directly
- Simplify `EvaluationSuite.run()` to delegate to `evaluate_suite`

## Example usage

```python
import sys
from typing import Any, Callable, Dict

from openai import OpenAI

import opik
from opik import EvaluationSuite, EvaluationSuiteResult
from opik.integrations.openai import track_openai

openai_client = track_openai(OpenAI())
opik_client = opik.Opik()

# ---------------------------------------------------------------------------
# Two prompt versions to compare
# ---------------------------------------------------------------------------

SYSTEM_PROMPT_V1 = """\
You are a Q&A assistant for Acme Cloud, a cloud hosting platform.
Answer user questions based on the provided documentation context.
If the answer is not in the context, say you don't know."""

SYSTEM_PROMPT_V2 = """\
You are a Q&A assistant for Acme Cloud, a cloud hosting platform.
Answer user questions based ONLY on the provided documentation context.
Rules:
- Never invent information not present in the context.
- If the answer is not in the context, say "I don't have that information."
- Keep answers concise (2-3 sentences max).
- When mentioning pricing, always include the currency and billing period."""


def make_task(system_prompt: str) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
    """Factory that returns a task function bound to a specific system prompt."""

    def task(item: Dict[str, Any]) -> Dict[str, Any]:
        question = item["question"]
        context = item.get("context", "")

        messages = [
            {"role": "system", "content": system_prompt},
            {
                "role": "user",
                "content": f"Question: {question}\n\nDocumentation context:\n{context}",
            },
        ]

        response = openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=messages,
            temperature=0.3,
            max_tokens=300,
        )

        return {
            "input": {"question": question, "context": context},
            "output": response.choices[0].message.content,
        }

    return task


def main() -> None:
    # Step 1: Create suite with baseline assertions and execution policy
    suite = opik_client.get_or_create_evaluation_suite(
        name="acme-cloud-qa-regression",
        description="Regression tests for Acme Cloud documentation Q&A bot",
        assertions=[
            "The response is grounded in the provided documentation context",
            "The response directly addresses the user's question",
        ],
        execution_policy={"runs_per_item": 1, "pass_threshold": 1},
        tags=["qa-bot", "rag", "regression"],
    )

    # Step 2: Batch add standard test cases (single dataset version)
    suite.add_items([
        {
            "data": {
                "question": "How do I create a new project?",
                "context": "Go to Dashboard > New Project. Enter a name and select a region.",
            },
            "description": "Basic how-to question",
        },
        {
            "data": {
                "question": "What are the pricing tiers?",
                "context": "Free ($0/month, 1GB), Pro ($29/month, 100GB), Enterprise (custom).",
            },
            "description": "Pricing question",
        },
    ])

    # Add a high-stakes item with custom assertions and stricter policy
    suite.add_item(
        data={
            "question": "How do I reset my API key?",
            "context": "Navigate to Settings > API Keys > Regenerate. Old key is invalidated.",
        },
        assertions=[
            "The response warns about the old key being invalidated",
            "The response advises to store the new key securely",
        ],
        execution_policy={"runs_per_item": 3, "pass_threshold": 2},
    )

    # Step 3: Run v1 — suite.run() prints pass/fail results automatically
    result_v1 = suite.run(
        task=make_task(SYSTEM_PROMPT_V1),
        experiment_name="acme-qa-v1-baseline",
        experiment_config={"prompt_version": "v1", "model": "gpt-4o-mini"},
        experiment_tags=["v1", "baseline"],
    )

    # Step 4: Evolve suite — tighten assertions, increase runs
    suite.update(
        assertions=[
            "The response is grounded in the provided documentation context",
            "The response directly addresses the user's question",
            "The response is concise and does not include unnecessary filler",
        ],
        execution_policy={"runs_per_item": 3, "pass_threshold": 2},
    )

    # Step 5: Run v2 with improved prompt
    result_v2 = suite.run(
        task=make_task(SYSTEM_PROMPT_V2),
        experiment_name="acme-qa-v2-improved",
        experiment_config={"prompt_version": "v2", "model": "gpt-4o-mini"},
        experiment_tags=["v2", "improved"],
    )

    # Step 6: Compare and CI gate
    print(f"v1 pass rate: {result_v1.pass_rate:.2f}")
    print(f"v2 pass rate: {result_v2.pass_rate:.2f}")

    if not result_v2.all_items_passed:
        print("CI GATE: FAILED")
        for item_id, item_result in result_v2.item_results.items():
            if not item_result.passed:
                print(f"  - {item_id}: {item_result.runs_passed}/{item_result.runs_total} "
                      f"(threshold: {item_result.pass_threshold})")
        sys.exit(1)
    else:
        print("CI GATE: PASSED")


if __name__ == "__main__":
    main()
```
image

## Change checklist
- [x] User facing
- [ ] Documentation update

## Issues

- [OPIK-4881](https://comet-ml.atlassian.net/browse/OPIK-4881) — `get_or_create_evaluation_suite` with new global assertions does not create a new version — assertions are dropped
- [OPIK-4882](https://comet-ml.atlassian.net/browse/OPIK-4882) — Each `eval_suite.add_item` call creates a new version — too noisy, no batch endpoint available
- [OPIK-4883](https://comet-ml.atlassian.net/browse/OPIK-4883) — Python SDK still contains references to `evaluators` (e.g. `get_evaluators`)
- [OPIK-4885](https://comet-ml.atlassian.net/browse/OPIK-4885) — Tags added to an evaluation suite do not appear in the UI
- [OPIK-4886](https://comet-ml.atlassian.net/browse/OPIK-4886) — Confusing API for `evaluation_suite.run` — expected input/output format is unclear, poor error messages
- [OPIK-4887](https://comet-ml.atlassian.net/browse/OPIK-4887) — SDK result display shows avg scores per assertion instead of overall pass rates
- [OPIK-4888](https://comet-ml.atlassian.net/browse/OPIK-4888) — Evaluation progress bar does not update while agent is running — feels stuck
- [OPIK-4889](https://comet-ml.atlassian.net/browse/OPIK-4889) — Experiment link generated by the SDK is broken
- [OPIK-4890](https://comet-ml.atlassian.net/browse/OPIK-4890) — `eval_suite.update(execution_policy=...)` raises ValueError about missing assertions/evaluators

## AI-WATERMARK

AI-WATERMARK: yes

- If yes:
  - Tools: Claude Code
  - Model(s): Claude Opus 4.6
  - Scope: All changes
  - Human verification: All code reviewed and tested by human

## Testing

- Unit tests: `cd sdks/python && python -m pytest tests/unit/api_objects/dataset/evaluation_suite/ tests/unit/evaluation/test_metrics_evaluator.py tests/unit/evaluation/suite_evaluators/test_llm_judge.py -v` (68 tests passing)
- E2E tests: `cd sdks/python && OPIK_URL_OVERRIDE=http://localhost:5173 python -m pytest tests/e2e/evaluation/test_evaluation_suite.py -v` (28 tests passing)
- Full evaluation unit tests: `cd sdks/python && python -m pytest tests/unit/evaluation/ tests/unit/api_objects/dataset/evaluation_suite/ -v` (421 tests passing)
- Linting: `cd sdks/python && pre-commit run --files <changed_files>` (all passing: ruff, ruff format, mypy, trailing whitespace, end of files)
- Manual: ran `evaluation_suite_advanced_example.py` end-to-end with timing and judge merging verified

## Documentation

[OPIK-4881]: https://comet-ml.atlassian.net/browse/OPIK-4881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py [hunk 1]
```diff
@@ -318,13 +375,13 @@ def run(
...
+        model: Optional[str] = None,
     ) -> suite_types.EvaluationSuiteResult:
         """
         Run the evaluation suite against a task function.
 
         The task function receives each test item's data dict and must return
-        a dict with "input" and "output" keys for the evaluators.
+        a dict with "input" and "output" keys.
 
         Args:
```
[the remark this answers is not in the document]
Comment (alexkuzmik) [reply sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py:405]: Backward compatibility is out of scope of this PR.

File: sdks/python/src/opik/api_objects/dataset/evaluation_suite/types.py [hunk 2]
```diff
@@ -3,12 +3,25 @@
 from __future__ import annotations
 
 import dataclasses
-from typing import Dict, List, Optional, TYPE_CHECKING
+from typing import Any, Dict, List, Optional, TYPE_CHECKING, TypedDict
+
+from typing_extensions import Required
```
[the remark this answers is not in the document]
Comment (alexkuzmik) [reply sdks/python/src/opik/api_objects/dataset/evaluation_suite/types.py:8]: type primitives are fine to be imported directly

File: sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py [hunk 3]
```diff
@@ -218,20 +222,18 @@ def get_execution_policy(self) -> execution_policy.ExecutionPolicy:
...
         Returns:
-            List of LLMJudge instances.
+            List of assertion strings.
         """
-        return self._dataset.get_evaluators(evaluator_model=evaluator_model)
+        evaluators = self._dataset.get_evaluators()
+        assertions: list[str] = []
+        for evaluator in evaluators:
+            assertions.extend(evaluator.assertions)
+        return assertions
```
[the remark this answers is not in the document]
Comment (alexkuzmik) [reply sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py:255]: Backward compatibility is out of scope of this PR.

File: sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py [hunk 4]
```diff
@@ -125,23 +122,20 @@ def get_items(
...
                         cfg = llm_judge_config.LLMJudgeConfig(**e.config)
-                        evaluator_objects.append(
-                            llm_judge.LLMJudge.from_config(
-                                cfg, init_kwargs={"model": evaluator_model}
-                            )
-                        )
+                        judge = llm_judge.LLMJudge.from_config(cfg)
+                        item_assertions.extend(judge.assertions)
 
             result.append(
```
[the remark this answers is not in the document]
Comment (alexkuzmik) [reply sdks/python/src/opik/api_objects/dataset/evaluation_suite/evaluation_suite.py:147]: Backward compatibility is out of scope of this PR.

File: sdks/python/tests/unit/evaluation/test_metrics_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
+from unittest.mock import MagicMock
+
+from opik.evaluation.engine.metrics_evaluator import build_metrics_evaluator
```
[the remark this answers is not in the document]
Comment (alexkuzmik) [reply sdks/python/tests/unit/evaluation/test_metrics_evaluator.py:3]: fine for tests

File: sdks/python/tests/unit/evaluation/test_metrics_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    def test_duplicate_assertions__deduplicated(self):
+        j1 = _make_judge(["A", "B"])
+        j2 = _make_judge(["B", "C"])
+
+        merged = llm_judge.LLMJudge.merged([j1, j2])
+
+        assert merged is not None
+        assert merged.assertions == ["A", "B", "C"]
+
+    def test_mismatched_settings__returns_none(self):
```
Comment (yaricom) [inline sdks/python/tests/unit/evaluation/test_metrics_evaluator.py:50]: Is there any warning that is needed to be checked as well?
